catroid/src/androidTest/java/org/catrobat/catroid/test/cloudmessaging/CloudMessagingTest.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2017 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.test.cloudmessaging;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+import android.support.test.InstrumentationRegistry;
+import android.test.AndroidTestCase;
+
+import org.catrobat.catroid.cloudmessaging.CloudMessaging;
+import org.catrobat.catroid.ui.SettingsActivity;
+import org.junit.After;
+import org.junit.Before;
+
+public class CloudMessagingTest extends AndroidTestCase {
+
+	private Context context;
+	private SharedPreferences.Editor editor;
+
+	@Before
+	public void setUp() {
+		context = InstrumentationRegistry.getTargetContext();
+		SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
+		editor = sharedPreferences.edit();
+		editor.clear();
+		editor.putBoolean(SettingsActivity.SETTINGS_CLOUD_NOTIFICATIONS, true);
+		editor.commit();
```

Comment:
```java
…
sharedPreferences = new FakeSharedPreferences();
context = mock(Context.class);
when(context.getSharedPreferences(anyString(), anyInt())).thenReturn(sharedPreferences);
…
```

If the FakeSharedPreferences and FakeSharedPreferences.Editor are partially implemented by simply using Maps, everything else can stay the same.

> as (used here) is the normal one

The difference is that you have to restore the SharedPreferences when using the "normal" method.
If a test crashes or you debug through a test and press the stop button, `tearDown` is never called and therefore the SharedPreferences aren't restored.

Maybe mentioning the [_normal_ one](https://github.com/Catrobat/Catroid/pull/2377/files#diff-c41b068c0724c4fe613f723e96219ba1R81) was a little confusing. I just thought that saving and restoring the (default) SharedPreferences can also be a feature of the FakeSharedPreferences :wink: